File: library/src/main/java/cn/cqray/android/dialog/amin/DialogAnimator.java
package cn.cqray.android.dialog.amin;

import android.animation.Animator;
import android.view.View;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import cn.cqray.android.anim.AnimatorBuilder;
import cn.cqray.android.anim.AnimatorListener;
import cn.cqray.android.anim.ViewAnimator;

/**
 * 动画框动画
 * @author Cqray
 */
public abstract class DialogAnimator {

    private ViewAnimator mAnimator;
    private List<AnimatorListener> mListeners = new ArrayList<>();

    public static void reset(@NonNull View view) {
        view.setAlpha(1);
        view.setScaleX(1);
        view.setScaleY(1);
        view.setTranslationX(0);
        view.setTranslationY(0);
        view.setRotation(0);
        view.setRotationY(0);
        view.setRotationX(0);
    }

    /**
     * 处理AnimatorBuilder
     * @param builder 动画构建器
     */
    public abstract void onHandle(@NonNull AnimatorBuilder builder);

    public void setTarget(View view) {
        mAnimator = ViewAnimator.playOn(view).convert();
    }

    public void addAnimatorListener(AnimatorListener listener) {
        mListeners.add(listener);
    }

    public void start() {
        AnimatorBuilder builder = mAnimator.getAnimatorBuilders().get(0);
        onHandle(builder);
        for (AnimatorListener listener : mListeners) {
            mAnimator.addAnimatorListener(listener);
        }
        mAnimator.start();
    }

    public void cancel() {
        if (mAnimator != null) {
            mAnimator.cancel();
        }
    }

    public boolean isRunning() {
        if (mAnimator != null) {
            return mAnimator.isRunning();
        }
        return false;
    }

    public void getDuration(ViewAnimator.Callback callback) {
        if (mAnimator != null) {
            mAnimator.getDuration(callback);
        }
    }

}

File: library/src/main/java/cn/cqray/android/dialog/amin/DialogAnimator.kt
package cn.cqray.android.dialog.amin

import android.view.View
import cn.cqray.android.anim.ViewAnimator
import cn.cqray.android.anim.listener.ViewAnimatorListener
import java.util.concurrent.atomic.AtomicReference

/**
 * 动画框动画
 * @author Cqray
 */
abstract class DialogAnimator {
    /** 目标视图 **/
    private val atomicView = AtomicReference<View>()

    /** 动画 **/
    private val atomicAnimator = AtomicReference<ViewAnimator>()

    /** 监听 **/
    private val listeners = mutableListOf<ViewAnimatorListener>()

    /** 动画是否正在运行 **/
    val isRunning get() = atomicAnimator.get()?.isRunning ?: false

    /** 使用的时间 **/
    val usedTime get() = atomicAnimator.get()?.getUsedTime() ?: 0

    /**
     * 当动画准备完毕
     * @param animator 动画
     */
    abstract fun onAnimatorPrepared(animator: ViewAnimator)

    fun setTarget(view: View) = atomicView.set(view)

    fun addAnimatorListener(listener: ViewAnimatorListener) = listeners.add(listener)

    fun start() {
        if (atomicView.get() == null) return
        atomicAnimator.set(ViewAnimator.with(atomicView.get()).also {
            listeners.forEach { listener -> it.addGlobalListener(listener) }
            it.duration(350L)
            onAnimatorPrepared(it)
            it.start()
        })
    }

    fun cancel() {
        atomicAnimator.get()?.cancel()
        atomicAnimator.set(null)
    }
}